### PR TITLE
Fix #117 issue, InRange SRP violation

### LINKED ISSUE #117: SRP Violation
I think InRange rule violates Single Responsibility Principle. And it will be useful if it is splited into InRange and Subset
### What steps will reproduce the problem?
```
$data1 = [
    'status' => 'active'
];
 
$data2 = [
    'status' => ['active']
];
 
$data3 = [
    'status' => ['active', 'not_active']
];
 
$rules = [
    'status' => [
        new \Yiisoft\Validator\Rule\InRange(['active', 'non_active', 'other_status'])
    ]
];
 
var_dump($validator->validate($data1, $rules)->isValid()); // true
var_dump($validator->validate($data2, $rules)->isValid()); // true
var_dump($validator->validate($data3, $rules)->isValid()); // true
```
### What is the expected result?
```
true
false
false
```
### What do you get instead?
```
true
true
true
```

### PR DESCRIPTION
**Pull request by kriptonuz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #117

### CODE REVIEW

**Comment by kriptonuz** [comment]
It's my forst pull request, Don't hit hard :)

File: src/Rule/Subset.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    private iterable $range;
+    /**
+     * @var bool whether the comparison is strict (both type and value must be the same)
+     */
+    private bool $strict = false;
+
+    private string $iterableMessage = 'Value must be iterable';
+
+    private string $subsetMessage = 'Value must be subset of...';
```
Comment (samdark) [inline src/Rule/Subset.php:25]:
```suggestion
    private string $subsetMessage = 'Values must be ones of {values}.';
```

File: src/Rule/Subset.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $result = new Result();
+
+        if (!is_iterable($value)) {
+            $result->addError($this->formatMessage($this->iterableMessage));
+            return $result;
+        }
+
+        if (!ArrayHelper::isSubset($value, $this->range, $this->strict)) {
+            $result->addError($this->formatMessage($this->subsetMessage));
```
Comment (samdark) [inline src/Rule/Subset.php:42]:
```suggestion
            $result->addError($this->formatMessage($this->subsetMessage, [
                'values' => '"' . implode('", "', $this->range) . '"',
            ]));
```

File: src/Rule/Subset.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\Rule;
+use Yiisoft\Validator\ValidationContext;
+
+class Subset extends Rule
+{
+    /**
+     * @var iterable
+     */
+    private iterable $range;
```
Comment (samdark) [inline src/Rule/Subset.php:17]:
```suggestion
    private iterable $values;
```

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by samdark** [comment]
Otherwise, good.

File: src/Rule/Subset.php (hunk 2 above)
Comment (kriptonuz) [reply src/Rule/Subset.php:42]: There will be error if $this->range is not list of strings

File: src/Rule/Subset.php [hunk 4]
```diff
@@ -38,8 +38,10 @@ protected function validateValue($value, ValidationContext $context = null): Res
             return $result;
         }
 
-        if (!ArrayHelper::isSubset($value, $this->range, $this->strict)) {
-            $result->addError($this->formatMessage($this->subsetMessage));
+        if (!ArrayHelper::isSubset($value, $this->values, $this->strict)) {
+            $result->addError($this->formatMessage($this->subsetMessage, [
+                'values' => '"' . implode('", "', $this->values) . '"',
```
Comment (samdark) [inline src/Rule/Subset.php:43]:
```suggestion
                'values' => '"' . implode('", "', iterator_to_array($this->values)) . '"',
```

File: src/Rule/Subset.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * @var bool whether the comparison is strict (both type and value must be the same)
+     */
+    private bool $strict = false;
+
+    private string $iterableMessage = 'Value must be iterable';
+
+    private string $subsetMessage = 'Value must be subset of...';
+
+    public function __construct(iterable $range)
```
Comment (kriptonuz) [inline src/Rule/Subset.php]: What if we maket its type array?

Comment (samdark) [reply src/Rule/Subset.php]: Why limiting use-cases?

**Comment by samdark** [comment]
👍

**Pull request merged**